File: runner/app/go/trickle/trickle_publisher.go
package main

import (
	"fmt"
	"io"
	"log/slog"
	"net/http"
	"sync"
)

// TricklePublisher represents a trickle streaming client
type TricklePublisher struct {
	baseURL     string
	streamName  string
	index       int          // Current index for segments
	writeLock   sync.Mutex   // Mutex to manage concurrent access
	pendingPost *pendingPost // Pre-initialized POST request
	contentType string
}

// pendingPost represents a pre-initialized POST request waiting for data
type pendingPost struct {
	index  int
	writer *io.PipeWriter
}

// NewTricklePublisher creates a new trickle stream client
func NewTricklePublisher(baseURL, streamName string) (*TricklePublisher, error) {
	c := &TricklePublisher{
		baseURL:     baseURL + "/realtime",
		streamName:  streamName,
		contentType: "video/MP2T",
	}
	p, err := c.preconnect()
	if err != nil {
		return nil, err
	}
	c.pendingPost = p

	return c, nil
}

// Acquire lock to manage access to pendingPost and index
// NB expects to have the lock already since we mutate the index
func (c *TricklePublisher) preconnect() (*pendingPost, error) {

	index := c.index
	url := fmt.Sprintf("%s/%s/%d", c.baseURL, c.streamName, index)

	pr, pw := io.Pipe()
	req, err := http.NewRequest("POST", url, pr)
	if err != nil {
		fmt.Printf("Failed to create request for segment %d: %v\n", index, err)
		return nil, err
	}
	req.Header.Set("Content-Type", c.contentType)

	// Start the POST request in a background goroutine
	// TODO error handling for these
	go func() {
		slog.Info("JOSH - initiailzing http client", "idx", index)
		resp, err := (&http.Client{}).Do(req) // prevent connection reuse
		if err != nil {
			slog.Info("Failed to complete POST for segment", "index", index, "err", err)
			return
		}
		body, err := io.ReadAll(resp.Body)
		if err != nil {
			slog.Info("Error reading body", "index", index, "err", err)
		}
		defer resp.Body.Close()

		if resp.StatusCode != http.StatusOK {
			slog.Info("Failed POST segment", "index", index, "status_code", resp.StatusCode, "msg", string(body))
		} else {
			slog.Info("Uploaded segment", "index", index)
		}
	}()

	c.index += 1
	return &pendingPost{
		writer: pw,
		index:  index,
	}, nil
}

func (c *TricklePublisher) Close() error {
	url := fmt.Sprintf("%s/%s", c.baseURL, c.streamName)
	req, err := http.NewRequest("DELETE", url, nil)
	if err != nil {
		return err
	}
	resp, err := (&http.Client{}).Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()
	if resp.StatusCode != http.StatusOK {
		body, _ := io.ReadAll(resp.Body)
		return fmt.Errorf("Failed to delete stream: %v - %s", resp.Status, string(body))
	}
	return nil
}

// Write sends data to the current segment, sets up the next segment concurrently, and blocks until completion
func (c *TricklePublisher) Write(data io.Reader) error {

	// Acquire lock to manage access to pendingPost and index
	c.writeLock.Lock()

	// Get the writer to use
	pp := c.pendingPost
	if pp == nil {
		p, err := c.preconnect()
		if err != nil {
			c.writeLock.Unlock()
			return err
		}
		pp = p
	}
	writer := pp.writer
	index := pp.index

	// Set up the next connection
	nextPost, err := c.preconnect()
	if err != nil {
		c.writeLock.Unlock()
		return err
	}
	c.pendingPost = nextPost

	// Now unlock so the copy does not block
	c.writeLock.Unlock()

	// Start streaming data to the current POST request
	n, err := io.Copy(writer, data)
	if err != nil {
		return fmt.Errorf("error streaming data to segment %d: %w", index, err)
	}

	slog.Info("Completed writing", "idx", index, "totalBytes", humanBytes(n))

	// Close the pipe writer to signal end of data for the current POST request
	if err := writer.Close(); err != nil {
		return fmt.Errorf("error closing writer for segment %d: %w", index, err)
	}

	return nil
}

func humanBytes(bytes int64) string {
	var unit int64 = 1024
	if bytes < unit {
		return fmt.Sprintf("%d B", bytes)
	}
	div, exp := unit, 0
	for n := bytes / unit; n >= unit; n /= unit {
		div *= unit
		exp++
	}
	return fmt.Sprintf("%.1f %cB", float64(bytes)/float64(div), "KMGTPE"[exp])
}

File: runner/app/live/trickle/trickle_publisher.py
import asyncio
import aiohttp
import logging
from contextlib import asynccontextmanager

class TricklePublisher:
    def __init__(self, url: str, mime_type: str):
        self.url = url
        self.mime_type = mime_type
        self.idx = 0  # Start index for POSTs
        self.next_writer = None
        self.lock = asyncio.Lock()  # Lock to manage concurrent access
        self.session = aiohttp.ClientSession()

    async def __aenter__(self):
        """Enter context manager."""
        return self

    async def __aexit__(self, exc_type, exc_value, traceback):
        """Exit context manager and close the session."""
        await self.close()

    def streamIdx(self):
        return f"{self.url}/{self.idx}"

    async def preconnect(self):
        """Preconnect to the server by initiating a POST request to the current index."""
        url = self.streamIdx()
        logging.info(f"Preconnecting to URL: {url}")
        try:
            # we will be incrementally writing data into this queue
            queue = asyncio.Queue()
            asyncio.create_task(self._run_post(url, queue))
            return queue
        except aiohttp.ClientError as e:
            logging.error(f"Failed to complete POST for {self.streamIdx()}: {e}")
            return None

    async def _run_post(self, url, queue):
        try:
            resp = await self.session.post(
                url,
                headers={'Connection': 'close', 'Content-Type': self.mime_type},
                data=self._stream_data(queue)
            )
            # TODO propagate errors?
            if resp.status != 200:
                body = await resp.text()
                logging.error(f"Trickle POST failed {self.streamIdx()}, status code: {resp.status}, msg: {body}")
        except Exception as e:
            logging.error(f"Trickle POST  exception {self.streamIdx()} - {e}")
        return None

    async def _stream_data(self, queue):
        """Stream data from the queue for the POST request."""
        while True:
            chunk = await queue.get()
            if chunk is None:  # Stop signal
                break
            yield chunk

    async def next(self):
        """Start or retrieve a pending POST request and preconnect for the next segment."""
        async with self.lock:
            if self.next_writer is None:
                logging.info(f"No pending connection, preconnecting {self.streamIdx()}...")
                self.next_writer = await self.preconnect()

            writer = self.next_writer
            self.next_writer = None

            # Set up the next POST in the background
            asyncio.create_task(self._preconnect_next_segment())

        return SegmentWriter(writer)

    async def _preconnect_next_segment(self):
        """Preconnect to the next POST in the background."""
        logging.info(f"Setting up next connection for {self.streamIdx()}")
        async with self.lock:
            if self.next_writer is not None:
                return
            self.idx += 1  # Increment the index for the next POST
            next_writer = await self.preconnect()
            if next_writer:
                self.next_writer = next_writer

    async def close(self):
        """Close the session when done."""
        logging.info(f"Closing {self.url}")
        if self.next_writer:
            s = SegmentWriter(self.next_writer)
            await s.close()
        await self.session.delete(self.url)
        await self.session.close()

class SegmentWriter:
    def __init__(self, queue: asyncio.Queue):
        self.queue = queue

    async def write(self, data):
        """Write data to the current segment."""
        await self.queue.put(data)

    async def close(self):
        """Ensure the request is properly closed when done."""
        await self.queue.put(None)  # Send None to signal end of data

    async def __aenter__(self):
        """Enter context manager."""
        return self

    async def __aexit__(self, exc_type, exc_value, traceback):
        """Exit context manager and close the connection."""
        await self.close()
